extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt

```diff
@@ -0,0 +1,42 @@
+package org.jboss.resteasy.reactive.server.runtime.kotlin
+
+import io.vertx.core.Context
+import kotlinx.coroutines.*
+import org.eclipse.microprofile.context.ManagedExecutor
+import javax.enterprise.context.ApplicationScoped
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Application wide coroutine scope. Should start and die with the rest of the application, along with child coroutines
+ * (structured concurrency).
+ *
+ * This should be the main interception point to place user specific [CoroutineContext.Element] objects. Things
+ * like activity ids, logger context propagation, tracing, exception handlers, etc.
+ */
+@ApplicationScoped
+class ApplicationCoroutineScope : CoroutineScope, AutoCloseable {
+    override val coroutineContext: CoroutineContext = SupervisorJob()
+
+    override fun close() {
+        coroutineContext.cancel()
+    }
+}
+
+/**
+ * Dispatches the coroutine in a worker thread from Vertx.
+ */
+class VertxDispatcher(private val vertxContext: Context) : CoroutineDispatcher() {
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        // todo sync context variables
+        vertxContext.runOnContext {
+            block.run()
+        }
+    }
+}
+
+class ExecutorDispatcher(private val managedExecutor: ManagedExecutor): CoroutineDispatcher() {
+    override fun dispatch(context: CoroutineContext, block: Runnable) {
+        managedExecutor.runAsync(block)
```

Comment:
Should we use `managedExecutor.execute(block)`? `runAsync` returns a `CompletableFuture`.

Reply:
Makes more sense. Will test and update if needed.

Reply:
>  This method should generally be exception-safe. An exception thrown from this method
 >  may leave the coroutines that use this dispatcher in the inconsistent and hard to debug state.

https://github.com/Kotlin/kotlinx.coroutines/blob/32336be8fa90818cff9f2ba3c36b3f7c89e079c4/kotlinx-coroutines-core/common/src/CoroutineDispatcher.kt#L70

As mentioned before, I would prefer asCoroutineDispatcher because the implementation of dispatch has subtle quirks, but that's my opinion.

Reply:
I understand your point. Will solve this issue because it is mainly a duplicate of the above.